cpp/src/arrow/ipc/reader.cc

```diff
@@ -1121,6 +1121,29 @@ class ARROW_EXPORT SelectiveIpcFileRecordBatchGenerator {
   int index_;
 };
 
+struct AtomicReadStats {
+  std::atomic<int64_t> num_messages{0};
+  std::atomic<int64_t> num_record_batches{0};
+  std::atomic<int64_t> num_dictionary_batches{0};
+  std::atomic<int64_t> num_dictionary_deltas{0};
+  std::atomic<int64_t> num_replaced_dictionaries{0};
+
+  /// \brief Capture a copy of the current counters
+  ///
+  /// It's possible to get inconsistent values.  For example, if
+  /// this method is called in the middle of a read you might have
+  /// a case where num_messages != num_record_batches + num_dictionary_batches
```

Comment:
This sounds a bit undesirable. Would it be too expensive to use a mutex instead?

Reply:
Does it make a difference that this is not new behavior?  We increment `num_messages` as soon as we figure out there is a new message.  We don't increment `num_record_batches` until we have confirmed that the new message is, in fact, a record batch.  This happens in an entirely different method altogether.

I can move things around to fix this but I also think this is mainly an internal counter used for unit tests and so I'm not very motivated to do so.  Mainly I made this comment because I was trying to figure out if this was an invariant we already maintained (we didn't) that I needed to keep consistent with the new approach.

Reply:
You're right, it is not new behaviour so we may not want to fix this anyway. OTOH, using a mutex instead of atomics should be an easy changes, a potential perf problem sounds like the only reasonable reason not to do it.

Reply:
In my limited understanding, since the values all likely reside in the same cache line, there's likely false sharing anyways right? So we may as well use the mutex.

Reply:
Looking at this further I'm more convinced this isn't needed.  For example, when reading rows, the number of messages will be incremented but the number of record batches will not.  So this isn't a very reliable invariant anyways.

I went ahead and removed the comment itself to avoid confusion in the future.

I'm still open to using a mutex instead of an atomic if we want to.  I only chose atomics for simplicity and not for performance.

There does appear to be a slight impact to performance for very simple IPC reads from buffers which is unfortunate:

Before:
```
ReadBuffer/num_cols:1/is_partial:0/real_time_mean         2372 ns         2372 ns          100 bytes_per_second=411.778G/s
```
After:
```
ReadBuffer/num_cols:1/is_partial:0/real_time_mean         2525 ns         2525 ns          100 bytes_per_second=386.79G/s
```

Reply:
> There does appear to be a slight impact to performance for very simple IPC reads from buffers which is unfortunate:

Don't be obsessed to microbenchmark :-)

Reply:
Thanks for the share.  We could maybe solve this particular problem with thread local stat counters that get accumulated when the stats are asked for but I agree with your earlier comment that we don't want to obsess over this too much until it becomes a problem for someone.